pkg/lease/manager.go

```diff
@@ -0,0 +1,205 @@
+}
+
+type manager struct {
+	client.Client
+	log logr.Logger
+}
+
+func (l *manager) CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error) {
+	return l.createOrGetExistingLease(ctx, node, duration, holderIdentity, namespace)
+}
+
+func (l *manager) UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error) {
+	return l.updateLease(ctx, node, lease, currentTime, leaseDuration, leaseDeadline, holderIdentity)
+}
+
+func (l *manager) InvalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error {
+	return l.invalidateLease(ctx, nodeName, leaseNamespace)
+}
+
+func NewManager(cl client.Client) Manager {
+	return NewManagerWithCustomLogger(cl, ctrl.Log.WithName("leaseManager"))
+
+}
+
+func NewManagerWithCustomLogger(cl client.Client, log logr.Logger) Manager {
+	return &manager{
+		Client: cl,
+		log:    log,
+	}
+}
+
+func (l *manager) createOrGetExistingLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, leaseNamespace string) (*coordv1.Lease, bool, error) {
+	owner := makeExpectedOwnerOfLease(node)
+	microTimeNow := metav1.NowMicro()
+
+	lease := &coordv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
+			Namespace:       leaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: pointer.Int32(int32(duration.Seconds())),
+			AcquireTime:          &microTimeNow,
+			RenewTime:            &microTimeNow,
+			LeaseTransitions:     pointer.Int32(0),
+		},
+	}
+
+	if err := l.Client.Create(ctx, lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+			nodeName := node.ObjectMeta.Name
+			key := apitypes.NamespacedName{Namespace: leaseNamespace, Name: nodeName}
+
+			if err := l.Client.Get(ctx, key, lease); err != nil {
+				return nil, false, err
+			}
+			return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false, nil
+}
+
+func (l *manager) updateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error) {
+	needUpdateLease := false
+	setAcquireAndLeaseTransitions := false
+	updateAlreadyOwnedLease := false
+
+	if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity == holderIdentity {
+		needUpdateLease, setAcquireAndLeaseTransitions = needUpdateOwnedLease(lease, *currentTime, leaseDeadline)
+		if needUpdateLease {
+			updateAlreadyOwnedLease = true
+
+			log.Infof("renew lease owned by nmo setAcquireTime=%t", setAcquireAndLeaseTransitions)
+
+		}
+	} else {
+		// can't update the lease if it is currently valid.
```

Comment:
nit: "can't take over the lease" would be clearer IMHO

Reply:
ping

Reply:
Yep, I agree